cypress/e2e/dashboard/chapters/chapter-administrator.cy.ts

```diff
@@ -0,0 +1,237 @@
+import { User } from '../../../../cypress.config';
+import { expectNoErrors } from '../../../support/util';
+
+const firstChapterId = 1;
+const secondChapterId = 2;
+
+describe('Chapter Administrator', () => {
+  let chapterRoles;
+  let instanceRoles;
+  let users;
+  before(() => {
+    cy.fixture('chapterRoles').then((fixture) => {
+      chapterRoles = fixture;
+    });
+    cy.fixture('instanceRoles').then((fixture) => {
+      instanceRoles = fixture;
+    });
+    cy.fixture('users').then((fixture) => {
+      users = fixture;
+    });
+  });
+  beforeEach(() => {
+    cy.task('seedDb');
+    cy.login(users.testUser.email);
+    cy.joinChapter(firstChapterId);
+  });
+
+  function confirmInstanceRole(email: string, expectedRole: string) {
+    cy.task<User>('getUser', email).then(({ instance_role: { name } }) => {
+      expect(name).to.eq(expectedRole);
+    });
+  }
+
+  function applyChapterRoleChanges(
+    userId: number,
+    changes: { chapterId: number; roleName: string }[],
+  ) {
+    for (const { chapterId, roleName } of changes) {
+      cy.changeChapterUserRole({ chapterId, userId, roleName }).then(
+        expectNoErrors,
+      );
+    }
+  }
+
+  describe('when promoted to administrator of chapter', () => {
+    it('instance member should get chapter_administrator instance role', () => {
+      cy.login();
+      cy.task<User>('getUser', users.testUser.email).then(
+        ({ id, instance_role }) => {
+          expect(instance_role.name).to.eq(instanceRoles.MEMBER);
+          applyChapterRoleChanges(id, [
+            { chapterId: firstChapterId, roleName: chapterRoles.ADMINISTRATOR },
+          ]);
+        },
+      );
+      confirmInstanceRole(
+        users.testUser.email,
+        instanceRoles.CHAPTER_ADMINISTRATOR,
+      );
```

Comment:
This is a common pattern.  They mostly take in an email, initial role, list of changes and expected roles.  It feels like it should be a function and then these tests should all become one or two calls to that function.

What do you reckon?  Apologies if I underestimated the complexity!

Reply:
You are right! It was hiding from me before.